gammapy/utils/coordinates/fov.py

```diff
@@ -1,7 +1,91 @@
+import numpy as np
+from astropy.coordinates import (
+    AltAz,
+    Angle,
+    BaseCoordinateFrame,
+    CoordinateAttribute,
+    DynamicMatrixTransform,
+    EarthLocationAttribute,
+    FunctionTransform,
+    RepresentationMapping,
+    SkyCoord,
+    SkyOffsetFrame,
+    TimeAttribute,
+    UnitSphericalRepresentation,
+    frame_transform_graph,
+)
+from astropy.coordinates.matrix_utilities import matrix_transpose, rotation_matrix
 
-__all__ = ["fov_to_sky", "sky_to_fov"]
+__all__ = ["FoVFrame", "fov_to_sky", "sky_to_fov"]
+
+reflect_lon_matrix = np.array([[1, 0, 0], [0, -1, 0], [0, 0, 1]])
+
+
+class FoVFrame(BaseCoordinateFrame):
+    """
+    FoV coordinate frame. Centered on `origin` and aligned on AltAz frame at `location` and `obstime`.
+
+    Longitudes are reversed.
+
+    Attributes
+    ----------
+    origin: astropy.coordinates.SkyCoord[AltAz]
+        Origin of this frame as a HorizonCoordinate
+    obstime: astropy.time.Time
+        Observation time
+    location: astropy.coordinates.EarthLocation
+        Location of the telescope
+    """
+
+    frame_specific_representation_info = {
+        UnitSphericalRepresentation: [
+            RepresentationMapping("lon", "fov_lon"),
+            RepresentationMapping("lat", "fov_lat"),
+        ]
+    }
+    default_representation = UnitSphericalRepresentation
+
+    origin = CoordinateAttribute(default=None, frame=AltAz)
+
+    obstime = TimeAttribute(default=None)
+    location = EarthLocationAttribute(default=None)
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        # make sure telescope coordinate is in range [-180°, 180°]
+        if isinstance(self._data, UnitSphericalRepresentation):
+            self._data.lon.wrap_angle = Angle(180, unit=u.deg)
+
+
+@frame_transform_graph.transform(FunctionTransform, FoVFrame, FoVFrame)
+def fov_to_fov(from_fov_coord, to_fov_frame):
+    """Transform between two skyoffset frames."""
+    intermediate_from = from_fov_coord.transform_to(from_fov_coord.origin)
+    intermediate_to = intermediate_from.transform_to(to_fov_frame.origin)
+    return intermediate_to.transform_to(to_fov_frame)
+
+
+@frame_transform_graph.transform(DynamicMatrixTransform, AltAz, FoVFrame)
+def altaz_to_fov(altaz_coord, fov_frame):
+    """Convert a reference coordinate to a sky offset frame."""
+    # Define rotation matrices along the position angle vector, and
+    # relative to the origin.
+    origin = fov_frame.origin.represent_as(UnitSphericalRepresentation)
+    mat1 = rotation_matrix(-origin.lat, "y")
+    mat2 = rotation_matrix(origin.lon, "z")
+
+    return reflect_lon_matrix @ mat1 @ mat2
+
+
+@frame_transform_graph.transform(DynamicMatrixTransform, FoVFrame, AltAz)
+def fov_to_altaz(fov_coord, altaz_frame):
+    """Convert an sky offset frame coordinate to the reference frame"""
+    # use the forward transform, but just invert it
+    mat = altaz_to_fov(altaz_frame, fov_coord)
```

Comment:
the first argument that is passed in a frame, where as in the function arguments the first is a coordinate. It is very confusing

Reply:
The comment above explains it. We use the inverse direction to compute the matrix, than invert the matrix.